src/prometheus/variant_counter.zig

```diff
@@ -0,0 +1,143 @@
+
+        pub fn observe(self: *Self, err: EnumOrError) void {
+            _ = self.counts[indexer.index(err)].fetchAdd(1, .monotonic);
+        }
+
+        pub fn getResult(metric: *Metric, _: std.mem.Allocator) Metric.Error!Metric.Result {
+            const self: *Self = @fieldParentPtr("metric", metric);
+            return .{ .variant_counter = .{
+                .counts = &self.counts,
+                .names = &names,
+            } };
+        }
+    };
+}
+
+pub const VariantCounts = struct {
+    counts: []const Atomic(u64),
+    names: []const []const u8,
+};
+
+/// Assigns a continuous sequence of integers starting at 0 to the
+/// variants. For example, if you have an variant set with 10 variants,
+/// it will assign the numbers 0-9 to each of the variants.
+///
+/// Enums usually already have numbers with this property, but it's
+/// not guaranteed. This is particularly useful for variants, whose
+/// integers are likely not continuous.
+///
+/// You can provide an variant to determine its index, or provide an
+/// index to get the respective variant.
+///
+/// This struct must be initialized at comptime, but it can be used
+/// at any time.
+const VariantIndexer = struct {
+    EnumOrError: type,
+    offset: u16,
+    index_to_int: []const u16,
+    int_to_index: []const u16,
+    len: usize,
+
+    const Self = @This();
+
+    pub fn init(comptime EnumOrError: type) Self {
+        const variants = switch (@typeInfo(EnumOrError)) {
+            .ErrorSet => |es| es.?,
+            .Enum => |e| e.fields,
+            else => @compileError(@typeName(EnumOrError) ++ " is neither error nor enum"),
+        };
+
+        // get min and max to determine array bounds and offset
+        var max: u16 = 0;
+        var min: u16 = std.math.maxInt(u16);
+        for (variants) |variant| {
+            const int = toInt(@field(EnumOrError, variant.name));
+            max = @max(max, int);
+            min = @min(min, int);
+        }
+        const offset = min;
+
+        // populate maps translating between the index and the variant's int representation
+        var init_index_to_int: [variants.len]u16 = undefined;
+        var init_int_to_index: [1 + max - min]u16 = undefined;
+        for (variants, 0..) |variant, i| {
+            const int = toInt(@field(EnumOrError, variant.name));
+            init_index_to_int[i] = int;
+            init_int_to_index[int - offset] = @intCast(i);
+        }
+        const index_to_int = init_index_to_int;
+        const int_to_index = init_int_to_index;
+
+        return .{
+            .EnumOrError = EnumOrError,
+            .offset = offset,
+            .index_to_int = &index_to_int,
+            .int_to_index = &int_to_index,
+            .len = index_to_int.len,
+        };
+    }
+
+    pub fn index(self: Self, err: self.EnumOrError) usize {
+        return self.int_to_index[toInt(err)];
+    }
+
+    pub fn get(self: Self, index_: usize) self.EnumOrError {
```

Comment:
is this used anywhere?

Reply:
I guess not. It's the inverse of `index` but not used so I can remove it c42bd45a1621a68db97e32d507b1529e3c1f9007